### ISSUE TITLE
Ghost batch normalisation

### ISSUE THREAD

**Issue by peterlee18** [manujosephv/pytorch_tabular#166]
Is it possible to add an option for ghost batch normalisation as well as batch normalisation? So one chooses either no batch norm, standard batch norm or ghost batch norm. There is an elegant implementation of ghost normalisation in the TabNet model.

This would allow the use of bigger batch sizes generally rather than just for the TabNet model - for example, in a standard FF network for fitting category feature embeddings.

If there is already an easy way to do this, then apologies. A pointer on how best to do this would be very helpful.

**Comment by manujosephv** [comment]
There is no easy way to include Ghost Batch Normalization right now. But, if GhostBatchNorm is a drop-in replacement to BatchNorm, then with a little bit of refactoring, one can parametrize the use of batchnorm and include GBN as well.

If you or anyone want to raise a PR for this, I'll be more than happy to merge it..

**Comment by peterlee18** [comment]
The code from TabNet for GBN is below

 

class GBN(nn.Module):

    """

    Ghost Batch Normalization

    https://arxiv.org/abs/1705.08741

 

    Parameters

    ----------

    input_dim : int

        Input dimension size

    virtual_batch_size : int, optional

        Virtual_batch_size, by default 128

    momentum : float, optional

        Momentum, by default 0.01

    """

 

    def __init__(self, input_dim, virtual_batch_size=128, momentum=0.01):

        """

        GBN class instantiation method.

        """

        super(GBN, self).__init__()

 

        self.input_dim = input_dim

        self.virtual_batch_size = virtual_batch_size

        self.bn = nn.BatchNorm1d(self.input_dim, momentum=momentum)

 

    def forward(self, x):

        """

        Forward propogation.

 

        Parameters

        ----------

        x : tensor

            Layer input.

 

        Returns

        -------

        tensor

            Input with normalised virtual batches.

        """

 

        chunks = x.chunk(int(np.ceil(x.shape[0] / self.virtual_batch_size)), 0)

        res = [self.bn(x_) for x_ in chunks]

**Comment by manujosephv** [comment]
Added with #346
